### PR TITLE
Cache args_as_names

### PR DESCRIPTION
**Pull request by Connormiha**
I cached `self.args_as_names()` before `forEach`. `args_as_names` quite heavy method.

### CODE REVIEW

File: lib/compress/index.js [hunk 1]
```diff
@@ -3915,9 +3915,11 @@ merge(Compressor.prototype, {
             if (vars_found > 0) {
                 // collect only vars which don't show up in self's arguments list
                 var defs = [];
-                vars.forEach(function(def, name) {
-                    if (self instanceof AST_Lambda
-                        && self.args_as_names().some((x) => x.name === def.name.name)) {
+                const args_as_names = self.args_as_names();
+                const isAstLambda = self instanceof AST_Lambda;
```
Comment (fabiosantoscode) [inline lib/compress/index.js]: Can you get rid of this check? It's redundant. This entire method should be on AST_Lambda anyway.

I was going to nitpick about the camelCase but the variable is not necessary in the first place.

**Comment by fabiosantoscode** [comment]
Thanks for the optimization. Just a small change :)

File: lib/compress/index.js (hunk 1 above)
Comment (Connormiha) [reply lib/compress/index.js]: Done

**Comment by fabiosantoscode** [comment]
Thanks!

**Comment by gdh1995** [comment]
I runs into an error of "self.args_as_names is not a function" when uglifying ES5 code, while ES6 code is safe to uglify.

Terser: 4.1.0, target code: https://github.com/gdh1995/vimium-c , command: `npm run chrome`.

The bundle.min.js reported such an error, and then I modified the terser/package.json to use `dist/bundle.js` but the `args_as_names` is still not a function, and it failed again even when I delayed the calling to `self.args_as_names()`.

**Comment by Connormiha** [comment]
Can you insert this code, before `forEach` please:
```js
if (typeod self.args_as_names !== 'function') {
    console.log(self);
    console.log(self.__proto__);
}
```

**Comment by gdh1995** [comment]
It printed an AST_Node instance, whose ctor is `[AST_Toplevel]`.

### `bundle.js`

``` javascript
if (typeof self.args_as_names !== 'function') {
  console.log("# =========== typeof self.args_as_names !== 'function': print self ==========");
  console.log(self);
  console.log("# =========== typeof self.args_as_names !== 'function': print __proto__ ==========");
  console.log(Object.prototype.toString.call(self.__proto__));
  console.log(self.__proto__);
}
```

### `npm run chrome`'s log

``` yaml
# =========== typeof self.args_as_names !== 'function': print self ==========
AST_Node {
  end: 
   AST_Token {
     raw: undefined,
     file: 'content\\vimium-c.js',
# ..................................... skipped .................................
# =========== typeof self.args_as_names !== 'function': print __proto__ ==========
[object Object]
AST_Node {
  CTOR: 
   { [Function: AST_Toplevel]
     BASE: 
# ..................................... skipped .................................
```

### Full log file

[vimium_c-npm-run-chrome.zip](https://github.com/terser-js/terser/files/3367599/vimium_c-npm-run-chrome.zip)

**Comment by fabiosantoscode** [comment]
I'm fixing this right now, thanks for the report. Could you please try to reproduce this crash with the least amount of code possible such that a test can be made for it? I would greatly appreciate it!

**Comment by gdh1995** [comment]
`input.js`:
```
var Foo;
var Bar;
```

config:
```
{
  "compress": {
    "hoist_vars": true
  }
}
```

**Comment by fabiosantoscode** [comment]
My god. It's incredible how there wasn't a test for that yet. Thank you very much! Go ahead and update terser, a new patch release is out :) thanks for the report and the test!

**Comment by gdh1995** [comment]
Thanks, but it crashes again.

``` bash
1$ rm -rf node_modules/
1$ rm -rf node_modules/
1$ la
total 5.0K
-rw-r--r-- 1 gdh1995 197613 46  7月  8 17:43 terserconfig.json
-rw-r--r-- 1 gdh1995 197613 18  7月  8 17:39 test-input.js
1$ ./node_modules/.bin/terser -c hoist_vars test-input.js
bash: ./node_modules/.bin/terser: No such file or directory
1$ npm install terser@4.1.1
npm WARN saveError ENOENT: no such file or directory, open 'R:\working\1\package.json'
npm WARN enoent ENOENT: no such file or directory, open 'R:\working\1\package.json'
npm WARN 1 No description
npm WARN 1 No repository field.
npm WARN 1 No README data
npm WARN 1 No license field.

+ terser@4.1.1
added 5 packages from 38 contributors and audited 6 packages in 1.154s
found 0 vulnerabilities

1$ ./node_modules/.bin/terser -c hoist_vars test-input.js
ERROR: t.args_as_names is not a function
    at AST_Toplevel.hoist_declarations (R:\working\1\node_modules\terser\dist\bundle.min.js:1:170326)
    at Mn.before (R:\working\1\node_modules\terser\dist\bundle.min.js:1:119140)
    at AST_Toplevel.transform (R:\working\1\node_modules\terser\dist\bundle.min.js:1:78599)
    at Mn.compress (R:\working\1\node_modules\terser\dist\bundle.min.js:1:118580)
    at Object.Bn [as minify] (R:\working\1\node_modules\terser\dist\bundle.min.js:1:315899)
    at run (R:\working\1\node_modules\terser\bin\uglifyjs:228:27)
    at Object.<anonymous> (R:\working\1\node_modules\terser\bin\uglifyjs:178:5)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
1$ cat test-input.js
var Foo;
var Bar;
```

**Comment by Connormiha** [comment]
@fabiosantoscode I fixed your patch
https://github.com/terser-js/terser/pull/395

**Comment by fabiosantoscode** [comment]
4.1.2 is out and this time there's a goddamn test lol

**Comment by gdh1995** [comment]
Thanks, then both the test demo and Vimium C's `npm run chrome` succeeded with the new PR.

**Pull request merged**